Fetching children of a translation unit results in an endless loop

When an exception occur while fetching children of a translation unit in the project explorer view:
![image](https://github.com/user-attachments/assets/173a77b1-c3e5-44f1-8387-19bc46ea699c)

The fetch will result in an endless call loop of `org.eclipse.cdt.lsp.internal.ui.navigator.SymbolsManager.refreshTreeContentFromLS(CompileUnit)`

```java
!ENTRY org.eclipse.lsp4e 4 0 2024-07-26 11:12:36.436
!MESSAGE java.io.UncheckedIOException: java.io.IOException: Cannot run program "C:\clangd\clangd_18.1.3\bi\clangd.exe" (in directory "C:\Users\blub\cdt-main\eclipse"): CreateProcess error=2, The system cannot find the file specified
!STACK 0
java.util.concurrent.CompletionException: java.io.UncheckedIOException: java.io.IOException: Cannot run program "C:\clangd\clangd_18.1.3\bi\clangd.exe" (in directory "C:\Users\blub\cdt-main\eclipse"): CreateProcess error=2, The system cannot find the file specified
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.io.UncheckedIOException: java.io.IOException: Cannot run program "C:\clangd\clangd_18.1.3\bi\clangd.exe" (in directory "C:\Users\blub\cdt-main\eclipse"): CreateProcess error=2, The system cannot find the file specified
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$0(LanguageServerWrapper.java:297)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	... 6 more
Caused by: java.io.IOException: Cannot run program "C:\clangd\clangd_18.1.3\bi\clangd.exe" (in directory "C:\Users\blub\cdt-main\eclipse"): CreateProcess error=2, The system cannot find the file specified
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1170)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1089)
	at org.eclipse.lsp4e.server.ProcessStreamConnectionProvider.start(ProcessStreamConnectionProvider.java:55)
	at org.eclipse.lsp4e.LoggingStreamConnectionProviderProxy.start(LoggingStreamConnectionProviderProxy.java:227)
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$0(LanguageServerWrapper.java:295)
	... 7 more
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
	at java.base/java.lang.ProcessImpl.create(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:500)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:159)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1126)
	... 11 more

!ENTRY org.eclipse.cdt.lsp 4 0 2024-07-26 11:12:36.437
!MESSAGE java.util.concurrent.CancellationException
!STACK 0
java.util.concurrent.ExecutionException: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at org.eclipse.cdt.lsp.internal.ui.navigator.SymbolsManager.refreshTreeContentFromLS(SymbolsManager.java:248)
	at org.eclipse.cdt.lsp.internal.ui.navigator.SymbolsManager.getCompileUnitElements(SymbolsManager.java:223)
	at org.eclipse.cdt.lsp.internal.ui.navigator.SymbolsManager.fetchDeferredChildren(SymbolsManager.java:129)
	at org.eclipse.ui.progress.DeferredTreeContentManager$1.run(DeferredTreeContentManager.java:215)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2510)
	at org.eclipse.lsp4e.LanguageServerWrapper.stop(LanguageServerWrapper.java:535)
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$9(LanguageServerWrapper.java:359)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

--- a/bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/internal/ui/navigator/SymbolsManager.java
+++ b/bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/internal/ui/navigator/SymbolsManager.java
@@ -239,6 +239,7 @@ public class SymbolsManager implements IDeferredWorkbenchAdapter {
 				temporaryLoadedDocument = true;
 			}
 			if (document != null) {
+				var isTimeoutException = new boolean[1];
 				CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> symbols;
 				final var params = new DocumentSymbolParams(LSPEclipseUtils.toTextDocumentIdentifier(document));
 				CompletableFuture<Optional<LanguageServerWrapper>> languageServer = LanguageServers
@@ -254,11 +255,13 @@ public class SymbolsManager implements IDeferredWorkbenchAdapter {
 					symbols = CompletableFuture.completedFuture(null);
 					if (e instanceof InterruptedException) {
 						Thread.currentThread().interrupt();
+					} else if (e instanceof TimeoutException) {
+						isTimeoutException[0] = true;
 					}
 				}
 				symbols.thenAcceptAsync(response -> {
 					compileUnit.symbolsModel.update(response);
-					compileUnit.isDirty = response == null; // reset dirty only when fetch was successful
+					compileUnit.isDirty = isTimeoutException[0]; // reset dirty only when no TimeoutException occurred
 				}).join();
 			} else {
 				temporaryLoadedDocument = false;

